// ng003-outillage-serveur/serveur.js
'use strict';

var jsonProxy = require('json-proxy');
var argv = require('yargs').argv;

var $ = require('gulp-load-plugins')();

/**
 * Ajoute les tâches Gulp pour lancer un serveur de dev
 * @param {Gulp} gulp
 * @param {Object} [options] les options
 */
module.exports = function (gulp, options) {

	/**
	 * Renvoie l'environnement demandé.
	 * @param env L'environnement à sélectionner pour proxifier les requêtes
	 */
	function findEnvironnement(env) {
		var selected = options.envs[env];
		console.log("Environnement sélectioné : " + env, selected);
		return jsonProxy.initialize(selected);
	}

	/**
	 * Démarre un serveur HTTP
	 * @param racine La racine à exposer.
	 * @param env L'environnement à sélectionner pour proxifier les requêtes
	 */
	function demarrerServeur(racine, env) {
		$.connect.server(
			{
				root: racine,
				port: 8888,
				livereload: true,
				middleware: function () {
					return [
						findEnvironnement(env)
					]
				}
			}
		);
	}


	/**
	 * Lance un serveur pour tester l'application en cours de développement.
	 * Dans votre navigateur, utiliser l'URL : "http://localhost:8888/"
	 */
	gulp.task('connect', function () {
		var env = argv.env || "defaut";
		demarrerServeur(['.', options.src], env);
	});

	/**
	 * Lance un serveur pour tester l'application en cours de développement
	 * en observant et rechargeant la page à chaque modification d'un fichier HTML ou Javascript.
	 * Dans votre navigateur, utiliser l'URL : "http://localhost:8888/"
	 */
	gulp.task('connect:watch', ['connect', 'watchHtml', 'watchJs']);

	/**
	 * Lance un serveur pour tester l'application distribuée.
	 * Dans votre navigateur, utiliser l'URL : "http://localhost:8888/"
	 */
	gulp.task('connect:dist', function () {
		var env = argv.env || "tuu";
		demarrerServeur(options.dist, env);
	});

	// --- Watch
	/**
	 * Observe les modifications des fichiers HTML et déclenche une action : rechargement du HTML dans le navigateur.
	 */
	gulp.task('watchHtml', function () {
		gulp.watch([options.src + '/**/*.html'], ['reloadHtml']);
	});

	/**
	 * Observe les modifications des fichiers Javascript et déclenche des actions :
	 * <ul>
	 *     <li>Exécution des tests unitaires</li>
	 *     <li>Rechargement du Javascript dans le navigateur</li>
	 * </ul>
	 */
	gulp.task('watchJs', function () {
		gulp.watch([options.src + '/**/*.js'], ['test', 'reloadJs']);
	});

	// --- Reload
	/**
	 * Recharge les fichiers HTML dans le navigateur.
	 */
	gulp.task('reloadHtml', function () {
		gulp.src(options.src + '/**/*.html')
			.pipe($.connect.reload());
	});


	/**
	 * Recharge les fichiers Javascript dans le navigateur.
	 */
	gulp.task('reloadJs', function () {
		gulp.src(options.src + '/**/*.js')
			.pipe($.connect.reload());
	});
};
